### PR TITLE
MPMorph Flows

### PR DESCRIPTION
**Pull request by BryantLi-BLI**
This PR seeks to port the original [MPMorph](https://github.com/materialsproject/mpmorph) workflows from atomate to atomate2. These workflows are used to generate equilibrated amorphous/crystalline/glassy structures with MD runs via AIMD.

Tagging in @Tinaatucsd, @mcgalcode, @mattmcdermott, @esoteric-ephemera, @vir-k01, @sivonxay because they have used it recently or are apart of the original development team.

Completed

- Volume equilibration workflow meant to equilibrate structure without NPT via a series of NVT runs (Equilibration + production volume)
- Code independent equilibration and production workflow, with optional quench methods
- VASP workflows used in MP papers below recreated in this PR:
  - [Hui Zheng et al.](https://arxiv.org/abs/2402.00177)
  - [Aykol and Persson 2018](https://pubs.acs.org/doi/full/10.1021/acsami.7b14868)
  - [Murat et al.](https://www.science.org/doi/full/10.1126/sciadv.aaq0148)
- There is an additional feature with this PR to incorporate MLFF MD runs with the same flow.
- Multi quench options available:
- Fast quench (double relax and static)
- Slow quench (series of MD runs to simulate cooling with different temperature profiles)
- Started to add tests for VASP
- MLFF tests are completed
- Pytests 
- Packmol conversion to pymatgen interface (used to generated initial amorphous structure)
- Slow quench temperature profiles

MISC

- Seeking feedback for code, missing features, usability

NOTE: Original PR closed and reopened due to some git commit history issues that caused large commit changes. This PR should have resolved the issue. Many thanks to @mcgalcode.

### CODE REVIEW

File: src/atomate2/vasp/flows/mpmorph.py
Comment (orionarcher) [inline src/atomate2/vasp/flows/mpmorph.py:47]: The differences between `BaseMPMorphVaspMDMaker`, `MPMorphVaspMDMaker`, and `MPMorphVaspMultiMDMaker` are not clear to me. Are all three needed? 

Also, is there a reason we have a `Base...` class here but no `Base...` class in the MLFF versions? I think minimizing the number of similar classes and standardizing the naming for the VASP and MLFF implementations would be helpful.

File: src/atomate2/vasp/flows/mpmorph.py
Comment (orionarcher) [inline src/atomate2/vasp/flows/mpmorph.py:80]: My comments on the force field maker re `quench_maker_kwargs`, setting `Maker` parameters in the `args`, and the need for a `_post_init_kwargs` argument also apply here.

**Review by orionarcher — COMMENTED** [review]
I am looking forward to using this once it's merged in! Thanks for all of this.

I left a number of comments and suggestions, mostly focused on parts of the interface I found confusing. Happy to say more if any of my comments are confusing or off-base. Y'all have much more context on the workflows than I do.

File: src/atomate2/common/jobs/mpmorph.py
Comment (esoteric-ephemera) [reply src/atomate2/common/jobs/mpmorph.py:263]: Will change the default to MP + cached (no API key needed). We're aiming to have the base workflow (no user kw/args) reproduce the MPMorph workflow used [here](https://arxiv.org/abs/2402.00177).

File: src/atomate2/forcefields/flows/mpmorph.py
Comment (esoteric-ephemera) [reply src/atomate2/forcefields/flows/mpmorph.py:77]: The idea was to ignore the temperature specified by the Maker / input set and use a common temperature in the initial equilibration and production runs. That gets around user error of setting different ensemble conditions in the various parts of the workflow

Then `_post_init_update` is needed on the VASP side of things because the temperature schedule and other relevant MD variables are fixed in the VASP input set. Similarly, for forcefields, the temperature is a Maker attr

Maybe there's a different paradigm that's clearer / makes more sense? Open to suggestions

File: src/atomate2/common/flows/mpmorph.py
Comment (BryantLi-BLI) [reply src/atomate2/common/flows/mpmorph.py:466]: So currently, MPMorph is only adaptable for VASP and MLFFMDMakers. The biggest issue is that MLFFMD and VASPMD takes in the kwargs for temperature different. The goal of the call_md_maker function is for it to be empty, and only when you call a specific version of SlowQuench (e.g SlowQuenchVaspMaker), it rebuilds the call_md_maker function properly. 

I envisioned the commons function to be empty, and when it needs to be adapted to other code, a new call_md_maker will be written to overwrite it, like SlowQuenchVaspMaker. I am open to suggestions on how to better handle this issue

File: src/atomate2/vasp/flows/mpmorph.py
Comment (BryantLi-BLI) [reply src/atomate2/vasp/flows/mpmorph.py:47]: `BaseMPMorphVaspMDMaker` is specifically for users who wants to customize their MPMorph settings. This enables them to adapt to specific MD makers, temperatures, n_steps, quench makers, convergence makers, etc. 

I wrote the other two as a clean wrapper over `BaseMPMorphVaspMDMaker`. `MPMorphVaspMDMaker` is meant for the generic, simplest case MPMorph, and `MPMorphVaspMultiMDMaker` are for systems where a long runtime is expected. I was hesitant putting them into the final repo. My only case for keeping them is user convenience. 

Also 100% agree with you the use of `Base...` is inconsistent. I will workshop this and come up with a better name for all the main, user friendly makers

File: src/atomate2/forcefields/flows/mpmorph.py
Comment (BryantLi-BLI) [reply src/atomate2/forcefields/flows/mpmorph.py:98]: I'm interested. I do believe it's possible to simplify `quench_maker_kwargs`. Currently I've written it this way to specifically separate FastQuenchMaker and SlowQuenchMaker. If I consolidate it into a giant quench maker, and within the kwargs it specifies how to switch between fast quench and slow quench that would be ideal. 

I can workshop this.

File: src/atomate2/forcefields/flows/mpmorph.py
Comment (BryantLi-BLI) [reply src/atomate2/forcefields/flows/mpmorph.py:122]: Ah...excellent observation. Will fix immediately. Thanks!

File: src/atomate2/common/flows/mpmorph.py
Comment (orionarcher) [reply src/atomate2/common/flows/mpmorph.py]: Ahh got it, that makes sense! I'd suggest raising a `NotImplementedError` here if the intent is for this to be reimplemented by child classes.

Comment (BryantLi-BLI) [reply src/atomate2/common/flows/mpmorph.py]: Yep! @esoteric-ephemera already fixed this!

File: src/atomate2/forcefields/flows/mpmorph.py
Comment (BryantLi-BLI) [reply src/atomate2/forcefields/flows/mpmorph.py]: So as a general response to the layout of `MPMorphMakers` across the forcefield and vasp flows, we wrote it with the post init function to make it user friendly. Typically, if a user needs to use MPMorph either on VASP or MLFF, they would have an expected temperature and number of steps they would run or their specific system (it varies a lot depending on the type of amorphous structure). Moreover, MLFF and VASP will be expected to have different ranges of n_steps because of the different scales it can be run at. So we wrote `MPMorphVaspMaker` and `MPMorphMLFFMaker` with the intention for it to be user friendly in adjusting these specific parameters. 

For the savvy `atomate2` user, I would recommend directly using `MPMorphMDMaker` from `atomate2.common.flows.mpmorph`, which is just the barebones logic for the MPMorph workflow. Using `MPMorphMDMaker` gives the user full flexibility on how to customize their calculations, adaptable to any VASP/MLFF input set requirements, and even beyond VASP and MLFF if required.

I do believe the overall question: "should we include such packaged function?" is important to address. I do worry if it causes too much codes to be built up in atomate2 over time. If that is true, then we can ax the pre-written user friendly makers, and we need to build it into the docs for the MPMorph flows instead. However, I do see the benefit for an average user to just import one maker instead of 4 different ones, each with their sets, just to use MPMorph. I am open to suggestions and feedback on this

File: src/atomate2/forcefields/flows/mpmorph.py
Comment (orionarcher) [reply src/atomate2/forcefields/flows/mpmorph.py]: Do you think that differentiating in the `quench_maker` itself would work? Something like:

```python
quench_maker = FastQuenchMaker(**quench_maker_kwargs)
MPMorphMLFFMaker(
    ...,
    quench_maker=quench_maker
)
```
for fast quench

```python
quench_maker = SlowQuenchMaker(**quench_maker_kwargs)
MPMorphMLFFMaker(
    ...,
    quench_maker=quench_maker
)
```
for slow quench

File: src/atomate2/forcefields/flows/mpmorph.py
Comment (orionarcher) [reply src/atomate2/forcefields/flows/mpmorph.py]: > we wrote it with the post init function to make it user friendly

This makes a lot of sense. Figuring out all of the parameters that need to be set for a maker is confusing and I think having easy and good defaults is ideal. I wonder if you could get the same effect with a user-friendly class method `MPMorphMLFFMaker.from_temp_and_steps` method (or some other `from_X` name) that automatically instantiates, `_post_init_updates`, and returns the makers. That would allow for a streamlined `__init__` signature and a beginner-friendly instantiation option.

```python
morph_maker = MPMorphMLFFMDMaker.from_temp_and_steps(
    temperature=300,
    steps_convergence=5000,
    steps_total_production=10000
```

```
morph_maker = MPMorphMLFFMDMaker(
    convergence_md_maker=...,
    quench_maker=...,
    production_md_maker=...,
)
```

It'd also allow for removing the `md_maker` arg. Currently, if `md_maker` and `convergence_md_maker` are both set, `md_maker` is silently ignored. Removing `md_maker` and instead having users set it within `convergence_md_maker` seems a bit cleaner.

> I do worry if it causes too much codes to be built up in atomate2 over time.

I definitely share this concern, there are a great many classes that are highly similar. I personally like the paradigm of having one class with multiple `from_XYZ` methods but there are tradeoffs there too.

File: src/atomate2/forcefields/flows/mpmorph.py
Comment (orionarcher) [reply src/atomate2/forcefields/flows/mpmorph.py:77]: As I responded above, I think a `MPMorphMLFFMaker.from_temp_and_steps` could be the solution here. Curious to hear what y'all think.

File: src/atomate2/forcefields/flows/mpmorph.py
Comment (esoteric-ephemera) [reply src/atomate2/forcefields/flows/mpmorph.py]: We removed `quench_maker_kwargs` entirely in favor of the user explicitly setting the `quench_maker` - no post init updates needed

File: src/atomate2/forcefields/flows/mpmorph.py
Comment (esoteric-ephemera) [reply src/atomate2/forcefields/flows/mpmorph.py]: We've added a `from_temperature_and_steps` method to the common, MLFF, and VASP `MPMorphMDMaker`'s. This lets the user input a common temperature. number of steps for the runs,  MDMaker / MD set, and then optionally a quench step.

The class itself is now instantiated by setting `convergence_md_maker`, `production_md_maker`, and `quench_maker` explicitly

File: src/atomate2/forcefields/flows/mpmorph.py
Comment (esoteric-ephemera) [reply src/atomate2/forcefields/flows/mpmorph.py]: Implemented as recommended - see above

File: src/atomate2/forcefields/flows/mpmorph.py
Comment (esoteric-ephemera) [reply src/atomate2/forcefields/flows/mpmorph.py:169]: I think so - the VASP and forcefield MD ensemble attributes ($N$, $V$, $T$, and $p$) are set in different locations (the `VaspInputSetGenerator` vs. just top-level `Maker` kwargs) so this might have to stay. Open to suggestions

File: src/atomate2/vasp/flows/mpmorph.py
Comment (esoteric-ephemera) [reply src/atomate2/vasp/flows/mpmorph.py]: We've cleaned up the names a lot to make these more consistent, let us know if this looks better!

File: src/atomate2/vasp/flows/mpmorph.py
Comment (esoteric-ephemera) [reply src/atomate2/vasp/flows/mpmorph.py]: Cleaned up with previous comments, I think

**Comment by BryantLi-BLI** [comment]
@utf @janosh @JaGeo: Should be ready for review!

File: src/atomate2/common/flows/mpmorph.py
Comment (orionarcher) [inline src/atomate2/common/flows/mpmorph.py:135]: Consider

```python
            # Fit EOS to the running volume and energy from previous calculations
            self.postprocessor.fit(working_outputs)
            working_outputs = dict(self.postprocessor.results)

            # breaks whole flow here if EOS is not fitted properly
            if working_outputs.get("V0") is None:
                return Response(output=working_outputs, stop_children=True)

            # Check if the number of calculations has reached the maximum
            n_calcs = len(working_outputs["relax"]["volume"])
            max_calcs = self.postprocessor.min_data_points + (self.max_attempts or 1000)
            max_calcs_reached = n_calcs >= max_calcs

            v_0 = working_outputs.get("V0")
            v_min = working_outputs.get("Vmin")
            v_max = working_outputs.get("Vmax")

            # return final structure if calculated volume is in interpolation regime
            if max_calcs_reached or (v_min <= v_0 <= v_max):
                final_structure = structure.copy()
                final_structure.scale_lattice(v_0)
                return Response(output=final_structure)

            # Else, if the extrapolated equilibrium volume is outside the range of
            # fitted volumes, scale appropriately
            v_ref = v_min if v_0 < v_min else v_max
            eps_0 = (v_0 / v_ref) ** (1 / 3) - 1.0

            linear_strains = [np.sign(eps_0) * (abs(eps_0) + self.min_strain)]
```

Comment (orionarcher) [reply src/atomate2/common/flows/mpmorph.py]: LGTM, I might save intermediate variables for the `if` statements but ultimately a matter of taste.

File: src/atomate2/common/flows/mpmorph.py
Comment (janosh) [inline src/atomate2/common/flows/mpmorph.py:215]: typo

File: src/atomate2/common/flows/mpmorph.py
Comment (janosh) [inline src/atomate2/common/flows/mpmorph.py:324]: is there any point in having generic `Makers` as defaults? nothing would happen unless you pass in explicit model/DFT `Makers`, no? in which case, maybe better not to set defaults at all?

File: src/atomate2/common/flows/mpmorph.py
Comment (janosh) [inline src/atomate2/common/flows/mpmorph.py:89]: this `make` looks like it could benefit from more checking of assumptions and raising helpful error messages if they fail.

e.g. 

```
       if isinstance(self.initial_strain, float | int):
          self.initial_strain = (
              -abs(self.initial_strain),
              abs(self.initial_strain),
          )
      elif not isinstance(self.initial_strain, tuple) or len(self.initial_strain) != 2:
         raise ValueError(f"initial_strain must be float | tuple[float, float], got {initial_strain}")
```

File: src/atomate2/common/flows/mpmorph.py
Comment (janosh) [inline src/atomate2/common/flows/mpmorph.py:190]: typo

File: src/atomate2/common/flows/mpmorph.py
Comment (janosh) [inline src/atomate2/common/flows/mpmorph.py:204]: maybe add a `__post_init__` to raise early if all (convergence_md_maker production_md_maker quench_maker) are left as `None`. based on the doc string sounds like maybe `production_md_maker` shouldn;t be allowed to be `None` anyway?

File: src/atomate2/common/flows/mpmorph.py
Comment (janosh) [inline src/atomate2/common/flows/mpmorph.py:273]: looks like this is missing an `@abstractmethod` decorator?

File: src/atomate2/common/flows/mpmorph.py
Comment (janosh) [inline src/atomate2/common/flows/mpmorph.py:476]: same here, should be marked @abstractmethod

File: src/atomate2/common/flows/mpmorph.py
Comment (janosh) [inline src/atomate2/common/flows/mpmorph.py:189]: the name `convergence_md_maker` seems a bit inconsistent. maybe call this `equilibrium_volume_maker`?

File: src/atomate2/common/flows/mpmorph.py
Comment (janosh) [inline src/atomate2/common/flows/mpmorph.py:193]: double space after :

File: src/atomate2/common/flows/mpmorph.py
Comment (janosh) [inline src/atomate2/common/flows/mpmorph.py:59]: `md_maker` is allowed to be `None` but looks like `make` will fail if it is? why not make this a required arg?

**Review by janosh — CHANGES_REQUESTED** [review]

File: src/atomate2/common/flows/mpmorph.py
Comment (esoteric-ephemera) [reply src/atomate2/common/flows/mpmorph.py:215]: Changed the docstr generally since it was confusing

File: src/atomate2/common/flows/mpmorph.py
Comment (esoteric-ephemera) [reply src/atomate2/common/flows/mpmorph.py:324]: The idea was probably to enforce that these are required, but the workflow itself will fail if `relax_maker` and `static_maker` are unspecified. Changed their defaults to `None`

File: src/atomate2/common/flows/mpmorph.py
Comment (esoteric-ephemera) [reply src/atomate2/common/flows/mpmorph.py:204]: Added a `__post_init__` to this and the base quench maker classes in the same file

File: src/atomate2/common/flows/mpmorph.py
Comment (esoteric-ephemera) [reply src/atomate2/common/flows/mpmorph.py:273]: Added this decorator to a few other places where I can see it's needed (eos common flows, ASE base classes, OpenMM base class)

File: src/atomate2/common/flows/mpmorph.py
Comment (esoteric-ephemera) [reply src/atomate2/common/flows/mpmorph.py:59]: See above, this has a `__post_init__` for checking this is not `None` now

Comment (janosh) [reply src/atomate2/common/flows/mpmorph.py:60]: i think adding a `__post_init__` can be considered boilerplate. if you make `md_maker` a required arg and the user doesn't pass it, you `dataclass` should raise a similar error message 

```
    def __post_init__(self) -> None:
        if self.md_maker is None:
            raise ValueError("You must specify `md_maker` to use this flow.")
```

Comment (esoteric-ephemera) [reply src/atomate2/common/flows/mpmorph.py:60]: Right - that's there already

Comment (janosh) [reply src/atomate2/common/flows/mpmorph.py:60]: i know, i'm suggesting removing it and changing

```diff
- md_maker: Maker | None = None
+ md_maker: Maker
```
to make `md_maker` a required arg

Comment (esoteric-ephemera) [reply src/atomate2/common/flows/mpmorph.py]: Ahhh yes sorry! Changed as requested, leaving in the `__post_init__`'s as well tho

File: src/atomate2/ase/jobs.py
Comment (janosh) [inline src/atomate2/ase/jobs.py:97]: was this change accidental? can't return `NotImplementedError` here, doesn't match the return type `Calculator`. should keep the `raise` and the `@abstractmethod` decorator

Comment (BryantLi-BLI) [reply src/atomate2/ase/jobs.py]: I swapped out `return` with `raise`.

I'm a bit confused on the `@abstractmethod` decorator. It seems that every forcefield that inherits the ASEMaker adapts their calculator as `@property`. Does this imply that all the calculators should be swapped from `@property` -> `@abstractmethod` ? or just the original base class where the calculator should be absent?

Comment (esoteric-ephemera) [reply src/atomate2/ase/jobs.py]: whoops typo, think @BryantLi-BLI  fixed that

Comment (janosh) [reply src/atomate2/ase/jobs.py]: > Does this imply that all the calculators should be swapped from @property -> @abstractmethod ? or just the original base class where the calculator should be absent?

just the base class should have the `@abstractmethod` decorator. but all the `calculator` methods should have the `@property` decorator

File: pyproject.toml
Comment (janosh) [inline pyproject.toml:142]: presumably this line is here to include `src/atomate2/common/jobs/mp_avg_vol.json.gz` with the PyPI package? probably best to pass the full file path to not auto-include future `json.gz` file in this dir by accident. but more importantly, that file is 650KB and needs to be downloaded by everyone who installs `atomate2` if we merge the current config, whether they use MPMorph flows or not. up to @utf to make a decision here but i would vote against adding that file to the package. better to auto-download it when the user first needs it with a link where to get it if user doesn't have an internet connection.

finally, i looked into that file. looks like it's partially redundant. at least the `oxi_state` and `count` keys can likely be removed?

Screenshot 2024-10-08 at 17 35 38

File: src/atomate2/ase/jobs.py
Comment (esoteric-ephemera) [reply src/atomate2/ase/jobs.py]: @BryantLi-BLI : that's my bad, the `AseMaker.calculator` should be both a property and abstractmethod. I accidentally removed this in [00d1434](https://github.com/materialsproject/atomate2/pull/938/commits/00d14347e6dc653b3aab0c4457b7bf7b14a21556) (was trying to see if there was a nice way to serialize ASE calculators without having to subclass AseMaker, and just forgot to restore a few things)

File: pyproject.toml
Comment (esoteric-ephemera) [reply pyproject.toml:142]: @janosh the oxidation states are I think ignored by default, but the counts are needed. If a given chemical environment doesn't exist in a database, the code looks through subspaces of that chemical environment and does a weighted average of structure volumes. The weighted average depends on the counts

I started rewriting this in parquet to jointly store them - might be better to put the parquet fiel on OpenData and pull down as needed

Comment (esoteric-ephemera) [reply pyproject.toml]: OK so for clarity: for the ICSD data, the user can either query from oxidation-state dependent or agnostic data, since ICSD includes manually-assigned charges. I pulled the MP data from summary rather than bonds, so the oxidation state info there is ignored.

That's a slight incongruity that I can change if requested

The counts are needed for the same reason given above (the code will sample subspaces of the total chemical space to perform a weighted averaging of reference volumes/atom if the total chemical space is not present in the dataset)

The data is now pulled / cached from figshare

Stuck with JSON since the parquet speed bump is probably not super relevant here (reading basically once rather than reading many times, and the user has the option to manually specify a DataFrame as input, if generating many random structures)

File: src/atomate2/common/flows/mpmorph.py
Comment (orionarcher) [inline src/atomate2/common/flows/mpmorph.py:148]: What if the return statements returned dictionaries instead of heterogenous outputs? e.g.

A successful return would look like:

```python
            if v0_in_range or max_attempts_reached:
                final_structure = structure.copy()
                final_structure.scale_lattice(v0)
                return {"structure": final_structure, "working outputs": working_outputs}
```

And a failing return:

```python
            if (v0 := working_outputs.get("V0")) is None:
                return Response(output={"working_outputs": working_outputs}, stop_children=True)

```

Advantages:

1. The API more closely mirrors the rest of atomate2 jobs, where we access the structure with `flow.output.structure`

2. By returning the working outputs at the end, there is better insight into how the job is working. I am currently finding the EOS fit is yielding unphysical densities when used with MLIPS. As is, it's a bit hard to collect the child jobs and inspect why the EOS fit is failing. If the working outputs were returned, that'd be very helpful.

3. The outputs can be evolved over time without breaking changes. If someone decided to include additional information in the output, that could be added as a new key:value pair without breaking anything.

Comment (esoteric-ephemera) [reply src/atomate2/common/flows/mpmorph.py]: Good suggestion! Added basically the same, but assigned the output prior to the `stress` and `energy` keys being popped from working outputs (that's the full EOS fit output, should be more useful for debugging)

**Comment by esoteric-ephemera** [comment]
Hey @utf, @janosh and @JaGeo: just want to ensure we meet the deadline for including new WFs. This should be ready for review whenever you have time!

**Review by janosh — APPROVED** [review]
based on the error message, sounds like the auto-download for `_DEFAULT_AVG_VOL_FILE` returns a 404 HTML page (or some other error code) instead of actual data. maybe mock the data instead of an actual download in CI? other than that, this looks great! 👍 

for context, @orionarcher has been running these workflows for a while. everything working as expected

**Comment by esoteric-ephemera** [comment]
@janosh good call and thanks to @orionarcher for testing this extensively!!

**Pull request merged**
